google/detectors/credentials/generic_weak_credential_detector/src/main/java/com/google/tsunami/plugins/detectors/credentials/genericweakcredentialdetector/testers/rstudio/RStudioCredentialTester.java

```diff
@@ -0,0 +1,271 @@
+    Document doc = Jsoup.parse(responseBody);
+    String title = doc.title();
+    String p =
+        doc.body().getElementsByTag("p").first().outerHtml().split("<p>")[1].split("</p>")[0];
+
+    if (title.contains(RSTUDIO_UNSUPPORTED_BROWSER_TITLE)
+        && p.contains(RSTUDIO_UNSUPPORTED_BROWSER_P)) {
+      logger.atInfo().log("Found RStudio endpoint");
+      return true;
+    } else {
+      return false;
+    }
+  }
+
+  @Override
+  public ImmutableList<TestCredential> testValidCredentials(
+      NetworkService networkService, List<TestCredential> credentials) {
+
+    return credentials.stream()
+        .filter(cred -> isRStudioAccessible(networkService, cred))
+        .collect(toImmutableList());
+  }
+
+  private boolean isRStudioAccessible(NetworkService networkService, TestCredential credential) {
+    var url = buildTargetUrl(networkService, "auth-public-key");
+    try {
+      logger.atInfo().log("Retrieving public key");
+      HttpResponse response = httpClient.send(get(url).withEmptyHeaders().build());
+      Optional<String> body = response.bodyString();
+      String exponent = body.get().split(":")[0];
+      String modulus = body.get().split(":")[1];
+
+      url = buildTargetUrl(networkService, "auth-do-sign-in");
+      logger.atInfo().log(
+          "url: %s, username: %s, password: %s",
+          url, credential.username(), credential.password().orElse(""));
+      response = sendRequestWithCredentials(url, credential, exponent, modulus);
+
+      if (response.headers().get("Set-Cookie").isPresent()) {
+        for (String s : response.headers().getAll("Set-Cookie")) {
+          if (s.contains("user-id=" + credential.username())) {
+            return true;
+          }
+        }
+      } else {
+        return false;
+      }
+    } catch (IOException
+        | NoSuchProviderException
+        | NoSuchAlgorithmException
+        | BadPaddingException
+        | IllegalBlockSizeException
+        | InvalidKeyException
+        | NoSuchPaddingException
+        | InvalidKeySpecException e) {
+      logger.atWarning().withCause(e).log("Unable to query '%s'.", url);
+    }
+    return false;
+  }
+
+  // This function base64 encodes provided cipertext string in hex.
+  private String hexToBase64(String hex) {
+    StringBuilder ret = new StringBuilder();
+
+    for (int i = 0; i + 3 <= hex.length(); i += 3) {
+      int c = Integer.parseInt(hex.substring(i, i + 3), 16);
+      ret.append(B64MAP.charAt(c >> 6)).append(B64MAP.charAt(c & 63));
+    }
+
+    int remaining = hex.length() % 3;
+
+    if (remaining == 1) {
+      int c = Integer.parseInt(hex.substring(hex.length() - 1), 16);
+      ret.append(B64MAP.charAt(c << 2)).append(B64MAP);
+    } else if (remaining == 2) {
+      int c = Integer.parseInt(hex.substring(hex.length() - 2), 16);
+      ret.append(B64MAP.charAt(c >> 2)).append(B64MAP.charAt((c & 3) << 4)).append(B64PAD);
+    }
+    ret.append(B64PAD);
+    return ret.toString();
+  }
+
```

Comment:
I haven't done this in Java, did a quick google search, does the [one-liner](https://codereview.stackexchange.com/a/111633) mentioned here work?

Reply:
During the plugin development i was looking for some shortcut to perform the conversion. Back then this was not working, but right now it is. Maybe it was related to other issues in the code. I've fixed this